base/base/bit_cast.h

```diff
@@ -13,6 +14,12 @@ template <typename To, typename From>
 std::decay_t<To> bit_cast(const From & from)
 {
     To res {};
-    memcpy(static_cast<void*>(&res), &from, std::min(sizeof(res), sizeof(from)));
+    if constexpr (std::endian::native == std::endian::little)
```

Comment:
I thought about this a bit. In general, I find usage of bit_cast with different source and target sizes quite dangerous depending on which types are casted to each other and how the target memory is initialized. For that reason, I converted all places where source and target are equally large to `std::bit_cast` (#42968) - the std-library bit_cast has that as a constraint.

Assume the source value is `0xAABBCCDD` (i.e. `sizeof(from) == 4`). Its BE representation is `0xAABBCCDD`, the LE representation is `0xDDCCBBAA`. Further assume, `sizeof(res) == 8` and that `res` is initially zeroed out.

With LE, the result after bit_cast is `0xDDCCBBAA00000000` (no change to the previous behavior).
With BE, we previously produced `0xAABBCCDD00000000` and we now produce `0x00000000AABBCCDD`.

If the result is interpreted as 8-byte number, then
- for LE, the result interpretation makes sense, i.e. the output value equals the input value
- for BE, the previous result made no sense (output != input) but the new one does.

I wonder how this logic worked previously. I guess it was broken but nobody noticed because we don't do tests on BE systems.

While you are at it, it would be nice though to add a `static_assert(sizeof(From) <= sizeof(To))` as a precondition. Doing so will also allow to rewrite l. 21 as `size_t offset = sizeof(res) - sizeof(to)`.

Reply:
\+ could you please add a comment to `bit_cast` which explains the memory layout and assumptions after casting (basically my previous comment). I fear that otherwise, it will not be obvious enough for users of `bit_cast`.

Reply:
Maybe we should add a static assert into CH bit_cast that fails if BE and sizeof(From) != sizeof(To) ?

Reply:
The point of CH-specific bit_cast is to allow for `sizeof(From) != sizeof(To)`, even if that is quite dangerous and needs the programmer to know what he/she is doing.

std::bit_cast is an alternative which has a static assert for `sizeof(From) == sizeof(To)`.

Reply:
Yep, I get the purpose of CH bit_cast, but my understanding is that currently it only works properly with sizeof(From) != sizeof(To) only on LE, but does something unexpected on BE. Is it true?

Reply:
"Working properly" is open for interpretation 😉 I think my example above where the target is initially all zero is the "safest" and most intuitive use case. But who knows which other tricky but valid use cases involving low level bit magic exist, e.g. I could imagine bit_cast-ing an integer into a IEEE 754 mantissa/exponent. Statically asserting sizeof(From) == sizeof(To) on BE would disallow such use cases.

Reply:
CI complaints about trailing whitespaces in the comment.

More importantly, the comment talks about "previous behavior" and "previously produced". This makes little sense in the context of a specific PR but less so for a code comment. Please replace the comment by the following:

```cpp
/// Assume the source value is 0xAABBCCDD, i.e. sizeof(from) == 4. Its big endian representation is
/// 0xAABBCCDD, its little endian representation is 0xDDCCBBAA. Further assume, sizeof(res) == 8
/// and that res is initially zeroed out.
/// With LE, the result after bit_cast will be 0xDDCCBBAA00000000 --> input value == output value.
/// With BE, the result after bit_cast will be 0x00000000AABBCCDD --> input value == output value.
```